pkg/controllers/provisioning/scheduling/topologygroup.go

```diff
@@ -15,54 +15,259 @@ limitations under the License.
+			if count < minCount {
+				minCount = count
+				minDomain = domain
+			}
+		}
+	}
+
+	// can we create new domains?
+	creatable := t.Key == v1.LabelHostname
+
+	// if it's a topology key that we can create a new domain for, there's effectively always a global min
+	// of zero
+	if creatable {
+		globalMin = 0
+	}
+
+	if len(t.domains) == 1 {
+		// There is only a single item in the domain.  If the domain is creatable we treat it as increasing skew since
+		// there is a hypothetical domain with a count of zero out there.  If it's not creatable, we treat it as
+		// not increasing skew since the domain has both the min/max counts and a skew of zero
+		if creatable {
+			return minDomain, t.maxCount + 1, true
+		}
+		return minDomain, 0, false
+	}
+
+	// none of the topology domains have any pods assigned, so we'll just be at
+	// a max-skew of 1 when we create something
+	if t.maxCount == 0 {
+		return minDomain, 1, true
+	}
+
+	// Calculate what the max skew will be if we chose the min domain.
+	maxSkew = t.maxCount - (minCount + 1)
+	if globalMin != minCount {
+		// if the global min is less than the count of pods in domains that match the node selector
+		// the max-skew is based on the global min as we can't change it
+		maxSkew = t.maxCount - globalMin
+		// the domain we're allowed to pick happens to be the maximum value, so by picking it we are increasing skew
+		// even more
+		if minCount == t.maxCount {
+			maxSkew++
+		}
+	}
+
+	// if max = min, skew would be negative since our new domain selection would be the new max
+	if maxSkew < 0 {
+		maxSkew = 1
+	}
+
+	// We need to know if we are increasing or decreasing skew.  If we are above the max-skew, but assigning this
+	// topology domain decreases skew, we should do it.
+	oldMaxSkew := t.maxCount - globalMin
+	increasingSkew = maxSkew > oldMaxSkew
+	return
+}
+
+func (t *TopologyGroup) AnyDomain(requirements v1alpha5.Requirements) string {
+	var requirement sets.Set
+	if requirements.Has(t.Key) {
+		requirement = requirements.Get(t.Key)
+	} else {
+		requirement = sets.NewComplementSet()
+	}
+	for domain := range t.domains {
+		if requirement.Has(domain) {
+			return domain
+		}
+	}
+	return ""
+}
+
+func (t *TopologyGroup) MaxNonZeroDomain(requirements v1alpha5.Requirements) string {
+	var requirement sets.Set
+	if requirements.Has(t.Key) {
+		requirement = requirements.Get(t.Key)
+	} else {
+		requirement = sets.NewComplementSet()
+	}
+	maxCount := int32(math.MinInt32)
+	var maxDomain string
+	for domain, count := range t.domains {
+		// we only consider domains that match the pod's node selectors
+		if requirement.Has(domain) {
+			if count > maxCount {
+				maxCount = count
+				maxDomain = domain
+			}
+		}
+	}
+	if maxCount == 0 {
+		return ""
+	}
+	return maxDomain
+}
+
+func (t *TopologyGroup) RegisterDomain(zone string) {
+	if _, ok := t.domains[zone]; !ok {
+		t.domains[zone] = 0
+	}
+}
+
+func (t *TopologyGroup) EmptyDomain(requirements v1alpha5.Requirements) string {
```

Comment:
Why does topology expose different interfaces for different topology types `EmptyDomain`, `MaxNonZeroDomain`, `AnyDomain`. It knows it's `ToplogyType`, so can't it just make the right decision with something simple like `topologyGroup.Next()`